frontend/middle-2/design.md

```diff
@@ -1,17 +1,46 @@
 #Базовые принципы проектирования
 
-* Что такое инкапсуляция? Какова мотивация для улучшения уровня инкапсуляции?
-* Что такое абстрагирование? Какие есть отличия между абстракцией, инкапсуляцией и скрытием информации?
-* Что такое coupling? Как инкапсуляция влияет на это?
-* Что такое cohesion? Является ли эта величина обратной величине coupling?
-* Что такое принцип DRY? Как принцип связан с cohesion?
-* Что такое принцип KISS?
-* Что такое принцип YAGNI? Как он соотносится с принципом KISS?
-* Что такое Separation of concerns?
-* Из-за чего возникает проблема impedance mismatch при обмене данными? Какие есть примеры возникновения проблемы в разных технологиях? Какие там используются решения?
-* Почему преждевременная оптимизация может легко обернуться проблемой, а не решением при проектировании систем?
-
+* Inversion of Control
+  * What is an IoC?
+  * What are advantages of framework with IoC over libraries?
+  * What are advantages of libraries without IoC over frameworks?
+  * Can we combine approach of using IoC and avoiding its within interface of the same module?
+* Dependency Injection
+  * What is DI?
+  * What are DI alternatives? What are advantages and disadvantages of each alternative?
```

Comment:
What are the alternatives to DI? What are the advantages and disadvantages of each alternative?